[IPCTestingAPI] JSGlobalObject may be destroyed when sending IPC messages during page teardown

charliewolfe opened the pull request:
#### de54ecba6b596ea666aa99205131b8cdb429a1e4
<pre>
[IPCTestingAPI] JSGlobalObject may be destroyed when sending IPC messages during page teardown
<a href="https://bugs.webkit.org/show_bug.cgi?id=270610">https://bugs.webkit.org/show_bug.cgi?id=270610</a>
<a href="https://rdar.apple.com/114871193">rdar://114871193</a>

Reviewed by Geoffrey Garen.

Some objects send IPC messages in their destructor. So, when the page is being torn down, it is possible
that the JSGlobalObject associated with the JSContextRef we are storing has already been destroyed. To
fix this, we should instead hold a weak reference to the global object and early return when it has been
destroyed.

* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSMessageListener::JSMessageListener):
(WebKit::IPCTestingAPI::JSMessageListener::didReceiveMessage):
(WebKit::IPCTestingAPI::JSMessageListener::willSendMessage):

Canonical link: <a href="https://commits.webkit.org/275792@main">https://commits.webkit.org/275792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/467bcf0bfd365cecc08071e4da9475319dc21880

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42810 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21831 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45211 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45423 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38934 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25556 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19211 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35420 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43383 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18890 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36842 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16384 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16502 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/866 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39020 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38241 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46935 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17637 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14531 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42156 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19262 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40795 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19426 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5800 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18892 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2851,7 +2852,7 @@ JSValueRef JSIPC::processTargets(JSContextRef context, JSObjectRef thisObject, J
     : m_jsIPC(jsIPC)
     , m_type(type)
-    , m_context(context)
+    , m_globalObject(JSC::jsCast<WebCore::JSDOMGlobalObject*>(toJS(context)))
```
Maybe it's better to pass the global object into the constructor instead.

geoffreygaren commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -349,7 +350,7 @@ class JSMessageListener final : public IPC::MessageObserver {
     WeakPtr<JSIPC> m_jsIPC;
     Type m_type;
-    JSContextRef m_context;
+    JSC::Weak<WebCore::JSDOMGlobalObject> m_globalObject;
```
This is way better than the status quo, so I'll say r+.

Did you consider storing a strong reference here? In the IPC testing infrastructure, is the JSMessageListener thingy reachable from the webpage? If so, strong would be a retain cycle. But if not, strong might be correct.

Is it a requirement that we use the same global object that the webpage uses? Did you consider just allocating our own global object, separate from the webpage? Or is there a specific need to use objects and types that come from the webpage when we do our IPC testing?

geoffreygaren replied:
The potential benefit of solving those follow-up issues is that you'd get testing coverage in these cases, instead of returning early.

charliewolfe replied:
Yes, preventing the global object from being garbage collected was what I tried to do first, but I realized it was wrong. These message listeners are reachable from the webpage.

I'm not completely sure if it would be okay to create our own global object. It does seem like some IPC decoding related things might break if the global object we used was separate from the webpage (see `JSMessageListener::jsDescriptionFromDecoder`), but I can't really explain what/why.

charliewolfe committed:
[IPCTestingAPI] JSGlobalObject may be destroyed when sending IPC messages during page teardown
https://bugs.webkit.org/show_bug.cgi?id=270610
rdar://114871193

Reviewed by Geoffrey Garen.

Some objects send IPC messages in their destructor. So, when the page is being torn down, it is possible
that the JSGlobalObject associated with the JSContextRef we are storing has already been destroyed. To
fix this, we should instead hold a weak reference to the global object and early return when it has been
destroyed.

* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSMessageListener::JSMessageListener):
(WebKit::IPCTestingAPI::JSMessageListener::didReceiveMessage):
(WebKit::IPCTestingAPI::JSMessageListener::willSendMessage):

Canonical link: https://commits.webkit.org/275792@main
